canonical/rt-tests-snap
# feat: Add `shared-memory` interface for cyclictest

## Summary
In its operation, `cyclictest` uses the POSIX shared memory API calls [shm_unlink(3)](https://linux.die.net/man/3/shm_unlink) and [shm_open(3)](https://man7.org/linux/man-pages/man3/shm_open.3.html) combined with [mmap(2)](https://man7.org/linux/man-pages/man2/mmap.2.html) and  [mlock(2)](https://man7.org/linux/man-pages/man2/mlockall.2.html) for efficient memory usage, for that, the `shared-memory` interface is needed in `rt-tests` snap.

* Add `shared-memory` interface to plugs
* Add additional test case for `cyclictest` 
* Add note on `README.md` mentioning `shared-memory` interface and linking to docs.

## Related issues
* Fixes #41

## Testing steps
* Download the snap from the `latest/edge/fix-shm-err` branch:
```bash
sudo snap install rt-tests --channel latest/edge/fix-shm-err
```
* Connect the interfaces: 
```bash
sudo snap connect rt-tests:process-control
sudo snap connect rt-tests:system-trace
sudo snap connect rt-tests:mount-observe
```
* Run `cyclictest`
```bash
sudo rt-tests.cyclictest --mlockall --smp  --priority=80 --interval=20 --loops=1000
```

### Results:

```console
lincoln@fenrir:~$ snap info rt-tests | grep tracking
tracking:     latest/edge/fix-shm-err
lincoln@fenrir:~$ snap connections rt-tests
Interface        Plug                                      Slot                        Notes
mount-observe    rt-tests:mount-observe                    :mount-observe              manual
posix-mq         -                                         rt-tests:rt-tests-posix-mq  -
process-control  rt-tests:process-control                  :process-control            manual
shared-memory    rt-tests:shared-memory                    :shared-memory              -
system-files     rt-tests:sys-kernel-debug-sched-features  :system-files               -
system-trace     rt-tests:system-trace                     :system-trace               manual
lincoln@fenrir:~$ sudo rt-tests.cyclictest --mlockall --smp  --priority=80 --interval=20 --loops=1000
WARN: open /dev/cpu_dma_latency: Operation not permitted
policy: fifo: loadavg: 12.47 12.51 12.57 14/3815 334004

T: 0 (333981) P:80 I:20 C:   1000 Min:      1 Act:    1 Avg:    1 Max:      12
T: 1 (333982) P:80 I:520 C:    148 Min:      2 Act:    2 Avg:    6 Max:      13
T: 2 (333983) P:80 I:1020 C:     70 Min:      2 Act:   11 Avg:    5 Max:      12
T: 3 (333984) P:80 I:1520 C:     46 Min:      2 Act:    4 Avg:    3 Max:      13
T: 4 (333985) P:80 I:2020 C:     33 Min:      2 Act:   12 Avg:    6 Max:      13
T: 5 (333986) P:80 I:2520 C:     24 Min:      2 Act:   12 Avg:    7 Max:      13
T: 6 (333987) P:80 I:3020 C:     19 Min:      2 Act:    2 Avg:    3 Max:       5
T: 7 (333988) P:80 I:3520 C:     15 Min:      2 Act:   13 Avg:    6 Max:      13
T: 8 (333989) P:80 I:4020 C:     12 Min:      5 Act:    6 Avg:   29 Max:     150
T: 9 (333990) P:80 I:4520 C:     10 Min:      3 Act:    3 Avg:    5 Max:      14
T:10 (333991) P:80 I:5020 C:      9 Min:      2 Act:    3 Avg:    7 Max:      12
T:11 (333992) P:80 I:5520 C:      7 Min:      2 Act:   12 Avg:    5 Max:      12
T:12 (333993) P:80 I:6020 C:      5 Min:      2 Act:    3 Avg:    2 Max:       3
T:13 (333994) P:80 I:6520 C:      5 Min:      2 Act:    3 Avg:    2 Max:       3
T:14 (333995) P:80 I:7020 C:      4 Min:     10 Act:   12 Avg:   11 Max:      12
T:15 (333996) P:80 I:7520 C:      3 Min:      2 Act:   12 Avg:    8 Max:      12
T:16 (333997) P:80 I:8020 C:      3 Min:      2 Act:   12 Avg:    8 Max:      12
T:17 (333998) P:80 I:8520 C:      2 Min:      3 Act:   12 Avg:    7 Max:      12
T:18 (333999) P:80 I:9020 C:      1 Min:      5 Act:    5 Avg:    5 Max:       5
T:19 (334000) P:80 I:9520 C:      1 Min:      3 Act:    3 Avg:    3 Max:       3
T:20 (334001) P:80 I:10020 C:      0 Min:1000000 Act:    0 Avg:    0 Max:       0
T:21 (334002) P:80 I:10520 C:      0 Min:1000000 Act:    0 Avg:    0 Max:       0
T:22 (334003) P:80 I:11020 C:      0 Min:1000000 Act:    0 Avg:    0 Max:       0
T:23 (334004) P:80 I:11520 C:      0 Min:1000000 Act:    0 Avg:    0 Max:       0
lincoln@fenrir:~$
```



**OBS**: Automated tests updated to include case where error happens.
